File: problems/0120_triangle/solution_top_down.ts
function minimumTotal(triangle: number[][]): number {
    // Top-down mirror of the bottom-up DP: best[i] = minimum path sum from
    // the apex down to column i of the current row.
    let best: number[] = [triangle[0][0]];
    for (const row of triangle.slice(1)) {
        // A cell descends from column i-1 or i of the row above, so both
        // ragged edge cells have a single parent.
        const nxt: number[] = [row[0] + best[0]];
        for (let i = 1; i < row.length - 1; i++) {
            nxt.push(row[i] + Math.min(best[i - 1], best[i]));
        }
        nxt.push(row[row.length - 1] + best[best.length - 1]);
        best = nxt;
    }
    // The answer is the cheapest cell on the final row.
    return Math.min(...best);
}

File: problems-originals/0101-0200/0120_triangle/solution_top_down.py
from typing import List, Optional


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        # Top-down mirror of the bottom-up DP: best[i] = minimum path sum
        # from the apex down to column i of the current row.
        best = list(triangle[0])
        for row in triangle[1:]:
            # Fresh row: a cell descends from column i-1 or i of the row
            # above, so both ragged edge cells have a single parent.
            nxt = [row[0] + best[0]]
            for i in range(1, len(row) - 1):
                nxt.append(row[i] + min(best[i - 1], best[i]))
            nxt.append(row[-1] + best[-1])
            best = nxt
        # The answer is the cheapest cell on the final row.
        return min(best)
